time-series/time-series-api/src/main/java/com/powsybl/timeseries/ast/UnaryOperation.java

```diff
@@ -56,26 +58,41 @@ public static UnaryOperation positive(NodeCalc child) {
     UnaryOperation(NodeCalc child, Operator operator) {
         super(child);
         this.operator = Objects.requireNonNull(operator);
+        this.hash = Objects.hash(child, NAME, operator);
     }
 
     public Operator getOperator() {
         return operator;
     }
 
     @Override
-    public <R, A> R accept(NodeCalcVisitor<R, A> visitor, A arg, int depth) {
+    public <R, A> R accept(NodeCalcVisitor<R, A> visitor, A arg, int depth, Map<Integer, R> cache) {
         if (depth < NodeCalcVisitors.RECURSION_THRESHOLD) {
-            NodeCalc child = visitor.iterate(this, arg);
-            R childValue = null;
-            if (child != null) {
-                childValue = child.accept(visitor, arg, depth + 1);
+            if (visitor instanceof NodeCalcEvaluator) {
+                if (cache.containsKey(this.hashCode())) {
```

Comment:
Using the hashcode for caching is not correct as we could have has code collisions. You should use System.identity() or even better an IdentityHashMap. 
Maybe also we should consider a different approach:
 - using a node number in the AST to have a cache based on an array direct access
 - even better and a totally different approach, modeling directly in the AST a NodeCalc reference at the building phase so that at eval phase only caches value where there is reference NodeCalc.